Rebase October Rain library on Laravel 6

bennothommo opened the pull request:
Please see https://github.com/octobercms/october/pull/4893 for more information.

This PR represents the current state of the Rain library rebase to Laravel 6.

Commit:
Removed now unused split package composer files

Commit:
Updated dependencies according to 5.6 upgrade guide

Commit:
BelongsTo method name changes in 5.8

Commit:
Updated dependencies to pull in Laravel 5.9

Commit:
Merge branch 'develop' into wip/laravel-5.9

Commit:
Merge remote-tracking branch 'origin/develop' into wip/laravel-5.9

Commit:
[Laravel 6] Bring code to a state where unit tests pass (#443)

Commit:
Drop `app:name` Artisan command.

It was removed in Laravel 6, and we have no real need for it as we have a custom namespace (and workspace) structure.

LukeTowers commented on composer.json:
```diff
@@ -12,38 +12,42 @@
-        "laravel/tinker": "~1.0"
+        "laravel/tinker": "^1.0",
+        "laravel/helpers": "^1.0",
+        "fideloper/proxy": "^4.0"
```
What is this dependency used for?

bennothommo replied:
@LukeTowers it's a trusted proxy library. It's required for Laravel 5.6, but it's only required by `laravel/laravel`, even though the framework uses it. So I had to make it a dependency for our library too.

LukeTowers replied:
@bennothommo could you point me to where it's required by the framework?

bennothommo replied:
@LukeTowers I was mistaken - for some reason, in my initial tests last year, it was erroring out if `fideloper/proxy` was missing. When I've done a fresh install, everything works now. The framework doesn't use it - only the `laravel/laravel` package does.

LukeTowers commented on src/Halcyon/MemoryRepository.php:
```diff
@@ -47,18 +47,18 @@ public function get($key, $default = null)
      * @return void
      */
-    public function put($key, $value, $minutes = null)
+    public function put($key, $value, $seconds = null)
```
Is there anywhere we need to ensure is updated to pass seconds instead of minutes to this?

bennothommo replied:
@LukeTowers very likely, but that's still a task to complete in the list.

LukeTowers commented on src/Support/Facade.php:
```diff
@@ -19,6 +19,7 @@ protected static function resolveFacadeInstance($name)
     {
         if (
             !is_object($name) &&
+            !is_null(static::$app) &&
```
Why was this required?

bennothommo replied:
@LukeTowers some unit tests fail without this, as no application is available (I believe only the plugin test cases initiate an application).

Commit:
Add e() helper into October helpers

In Laravel 5.6, they changed $doubleEncode to "true" by default. As a LOT of functionality in OCMS has no $doubleEncode specified, so this is resulting in a lot of double-encoded content (ie. content for code editors or rich editors).

Our override brings the default back to "false" to allow all current functionality to work without change.

Commit:
Remove extra space

Commit:
Implement Laravel helpers in October helpers, remove trusted proxy lib

Commit:
Drop our custom log provider.

The two methods it provides are not being used anymore and simply seem to define the location of the log, which is now done in "config/logging.php" and handled by Laravel.

Commit:
Drop unused Exception use case

Commit:
Add HasOneThrough relation. Update comments for HasManyThrough.

Commit:
Retain support for Mandrill and SparkPost mail transports

Commit:
Fix code smells

Commit:
Force jenssegers/date to minimum version of 3.5 to support Carbon 2

Commit:
Convert cache expiry for Builder objects to DateTime

Commit:
Update default rules for base User model

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Added a subsplit script for the laravel 6 upgrade.

Commit:
Retain setGlobalTo method name for BC.

Fixes https://github.com/octobercms/october/issues/4915

Commit:
Replace setGlobalTo reference.

Since the method is protected, it would be a better fix to replace our usage and keep parity with Laravel.

Replaces 13efd83fc9f10ff21d2b174ed6a5c033e9914699

Commit:
Adds support for namespace mapping in ClassLoader.

This is needed for tests, which will use a directory structure similar to plugins and modules, but has a different namespace structure.

This change should be backwards-compatible with any other uses of the ClassLoader.

Commit:
Revert "Adds support for namespace mapping in ClassLoader."

This reverts commit 8d5cf3c2072d07232b58288d9a9031d2ff2dd0cd.

Commit:
Namespace library tests

Commit:
Flush models on initial boot of DB/Halcyon services.

Should resolve some issues in testing where models are not properly reset to a blank state.

Refs: https://github.com/rainlab/translate-plugin/issues/546, https://github.com/rainlab/translate-plugin/pull/509

Commit:
Add clearDuplicate method for Halycon models.

Acts similar to the same method in the database models.

Commit:
Fix null issue if cache manager is not specified in Halcyon model.

LukeTowers commented on composer.json:
```diff
@@ -12,38 +12,40 @@
+        "illuminate/validation": "^6.0",
+        "illuminate/view": "^6.0",
+        "illuminate/cache": "^6.0",
+        "meyfa/phpunit-assert-gd": "^2.0.0",
```
Do you mind grouping all the PHPUnit deps next to each other? Additionally, why do we include all these laravel libraries manually if we include laravel/framework in the main project?

bennothommo replied:
@LukeTowers my understanding is that the framework libraries are there so the library can be tested in isolation.

LukeTowers replied:
@bennothommo fair enough, wasn't sure if the main project was being pulled in to run these tests or not

bennothommo replied:
@LukeTowers not with the automated tests on this repo, no. I think @daftspunk has mentioned in the past that the library is sometimes used externally to October.

Commit:
Restore SCSS dependency for SASS / SCSS compilation support

Commit:
Merge branch 'wip/laravel-6' of https://github.com/octobercms/library into wip/laravel-6

Commit:
Replace JSMin with JSqueeze

Refs: https://github.com/octobercms/october/commit/627c0aa6431ad67c6575851b85630856eb5448d8

Commit:
Update Assetic SCSS filter reference

Commit:
Add key resolver to allow signed URLs to work

Commit:
Re-implement support for Rackspace

Commit:
Fix code smell

Commit:
Merge remote-tracking branch 'origin/develop' into wip/laravel-6

daftspunk committed:
Configure default logger

This accounts for application instances that do not have the logger.php configuration file, these will fall back to the default of using the single file logger

daftspunk committed:
Code smell

daftspunk committed:
Switch back to October LogServiceProvider

Commit:
Add Redis provider override

Laravel 6 is defaulting to `phpredis` as the client of choice going forward. To maintain BC, we'll override the provider and default to `predis`.

Users who wish to use `phpredis` need only set the `database.redis.client` config value to `phpredis`.

daftspunk committed:
Drop namespaces on tests

daftspunk committed:
Switch to wikimedia/less

Implement basic assetic fork - This contains no use of unsafe filters

Move back to simple JSMin filter

daftspunk committed:
Use Input instead of Request

There is a lot of documentation out there that refers to "Input", since they are interchangable it is ok to leave it as it was

Introduce router helpers

daftspunk committed:
Fixes code smells

daftspunk committed:
Remove non-existent CS ignore

Commit:
Add eloquent.retrieved event that was added in Laravel 5.5.2.

Refs: https://github.com/laravel/framework/commit/bc9ae14af320e57238bb5d631868993c82120c6c

Commit:
Bring event dispatcher more in line with Laravel's dispatcher.

This brings October's event dispatcher more in line with Laravel's by changing the parameters sent to wild card listeners (introduced in https://github.com/laravel/framework/commit/dbbfc62beff1625b0d45bbf39650d047555cf4fa but accidentally overlooked in the 5.5 update release).

This changes wildcard event listener signatures from function (...$params) {} to function ($event, $params) {}. Note that this change basically obsolutes the Event::firing() functionality, so a decision to keep or remove that will be required as well.

Additionally, this adds in the wildcard event listener cache added in https://github.com/laravel/framework/commit/c86efa6caade8aa6a1507012ccb9522eb554dfe1

Commit:
Removed no longer valid test

rajakhoury committed:
Fix Assetic SCSS use statement (#468)

Commit:
Change some other Leafo class namespaces to ScssPhp

rajakhoury committed:
Fix code sniffer issues (#469)

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Return result of flush method

Since Laravel 5.7 the flush() method has been expected to return a result for the sake of the `cache:clear` command. Docblocks have said a result would be returned for longer than that though, so this is just a bit of housekeeping: https://github.com/laravel/framework/commit/ad670c8423a5fdfc929644f63db1b7ca39a8ba59

rajakhoury committed:
Use correct method for marking providers as deferred (#470)

Laravel now uses an interface to mark providers as deferred, based on the following release notes: https://laravel.com/docs/5.8/upgrade#deferred-service-providers

jacobdekeizer committed:
Add required PHP extensions to composer.json (#476)

More clearly documents the project's requirements.

Commit:
Increase range of data that the field parser will accept

Commit:
Merge branch 'develop' into wip/laravel-6

Samuell1 committed:
Add hasManyThrough, hasOneThrough relation definition in scaffold (#459)

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Use path.config instance when getting config path

Laravel committed https://github.com/laravel/framework/commit/021b9158a1dd11bb32ee358de18eff06e883fa13 ostensibly to remove double path checking as a micro-optimisation. It also prevents the config path being changed dynamically at runtime, preventing our tests from working, as it was forced to config path specified in the Application.

Commit:
Merge branch 'develop' into wip/laravel-6

# Conflicts:
#	composer.json

Commit:
Merge branch 'develop' into wip/laravel-6

# Conflicts:
#	composer.json

Commit:
Pass current asset path down to children to use as new root path. (#484)

Previously, this method would pass the same loadPath to all assets down the hierarchy, which would result in any relative paths not being able to be found. This change will take the current directory of the current asset and use that as a base for all child assets.

mjauvin committed:
Re-add Assetic LessphpFilter (#485)

Adds the missing Asset\Filter\LessphpFilter

Co-authored-by: Ben Thomson <ben@abweb.com.au>

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Fix compatibility with Laravel 6.

Related: https://github.com/laravel/framework/pull/31677 & https://github.com/jenssegers/laravel-mongodb/pull/1979

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Required for scss-php

Commit:
Updated required extensions.

Both of these should be present by default on most installations, but just being explicit

Commit:
Add required extensions from TijsVerkoyen\CssToInlineStyles.

Used by System\Classes\MailManager

Commit:
Document required PDO extensions for connecting to various types of databases

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Merge remote-tracking branch 'origin/develop' into wip/laravel-6

Commit:
Make build scripts executable

Commit:
Wrap Laravel Validation service.

Laravel 5.8 introduced a new email validation library, with a more permissive default filter (supposedly, following the RFC). To maintain the expected functionality of October, we have reverted the default back to the 'filter' method, which uses PHP's inbuilt 'filter_var' method.

Fixes https://github.com/octobercms/october/issues/5070

Commit:
Add unit test for email validation, update Halcyon Model test

Commit:
Fix library test for email validation - seems facades aren't available.

Commit:
Sigh, seems there isn't an application either in testing

multiwebinc committed:
Update Laravel Tinker to 2.0 (#494)

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Improve stability of L6 Translation service

Reimplemented removed trans & transChoice methods and implemented https://github.com/laravel/framework/commit/697b898a1c89881c91af83ecc4493fa681e2aa38

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Add new `create:theme $code` scaffold command

Commit:
Fix minor visual issue with scaffolded themes

Commit:
Merge develop into wip/laravel-6

Commit:
Add array_undot helper method

Commit:
Add a courtesy link

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Merge remote-tracking branch 'origin/develop' into wip/laravel-6

Commit:
Fix Encryptable unit test for PHPUnit 8

Commit:
Allow deep symlinks when translating paths to local URLs (#491)

This extends the work done by @LukeTowers in https://github.com/octobercms/library/commit/a88b5fd3880edeac0c2bb3bf9e5297c508300055 by doing a search (once) for all symlinks in the base directory (or in any subdirectory if `develop.allowDeepSymlinks` is enabled)

It will create an array of sources and targets in order to translate these to URLs.

In addition, this will respect the cms.restrictBaseDir config value, in that symlinks that link to folders outside the base directory will not be recorded if restricted to the base directory, but symlinks that remain within the base directory will still work.

Commit:
Add L6 as a dep to the library to prevent people with older Laravel versions from accidentally pulling int he L6 update

Commit:
Revert the min password length requirement to 4 like it was originally

LukeTowers commented on src/Database/Concerns/HasRelationships.php:
```diff
@@ -299,6 +321,7 @@ protected function handleRelation($relationName)
                 $relationObj = $this->$relationType($relation[0], $relation['public'], $relation['key'], $relationName);
                 break;
 
+            case 'hasOneThrough':
```
Was this added to the database docs?

bennothommo replied:
@LukeTowers yep :) - https://github.com/octobercms/docs/blob/wip/laravel-6/database-relations.md#has-one-through

LukeTowers replied:
Excellent, thanks!

bennothommo commented on src/Auth/Models/User.php:
```diff
@@ -25,8 +25,8 @@ class User extends Model implements \Illuminate\Contracts\Auth\Authenticatable
         'email' => 'required|between:3,255|email|unique:users',
-        'password' => 'required:create|min:2|confirmed',
-        'password_confirmation' => 'required_with:password|min:2'
+        'password' => 'required:create|min:4|confirmed',
+        'password_confirmation' => 'required_with:password|min:4'
```
@LukeTowers I'm not quite sure the reasoning that @daftspunk changed this back to `2` (I presume for backwards compatibility), but if we are going to change it, I personally would rather this be `min:6` or `min:8` at the least. We should be following best practices for password selection, and 4 character passwords are too easy to brute-force in this day and age. (I understand we have throttling, but it can be disabled).

LukeTowers replied:
@bennothommo I checked the git history and the value was originally 4, so I don't know why @daftspunk felt the need to bump it down to 2. I would also prefer the minimum be set to 8, but I believe we had a conversation about this a while ago and it was decided that people could always just bump up the rules themselves. Additionally, this is somewhat tied into the autogeneration of admin passwords PR, was that one ever finished?

bennothommo replied:
@LukeTowers I do recall that conversation. On further consideration though, I feel if anything we should be bumping it high and forcing people to work to make the value lower, so that they are fully responsible for the consequences.

Re: auto-generation of admin passwords, it's more or less [finished](https://github.com/octobercms/october/pull/4866), but on hold until after L6.

LukeTowers replied:
I'll have another convo with @daftspunk about it and get back to you

LukeTowers replied:
Spoke to Sam and he re-iterated that
> it's best to leave it as 4 otherwise it's a breaking change. Password policies are a complex topic and so it's best not to have any strong opinions and leave it up to a plugin or attend to the entire problem holistically
> there are legit applications that use short passwords
> Namely demo apps and temporary passwords
> A config item would probably appease whatever is driving this. Default without a breakage

So we could add a new config item in `auth.minPasswordLength` that defaults to `8` in the file but in the Config::get() it defaults to the existing 4. That way existing installs won't break, and new ones will default to 8.

Also we might want to look at namespacing the auth config under auth.backend instead to better support any future changes bringing us closer to Laravel's auth system.

Commit:
Properly support DateTime instances as well as minutes for the QueryBuilder->remember() method.

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Add "resolve_path" helper method (#509)

Adds a resolve_path() method that operates very similar to realpath() except in that it will allow missing files and folders. It will resolve relative directory markers and expand any symlinks encountered.

Refs: https://github.com/octobercms/october/pull/5229

bennothommo committed:
Fix canonical path tests when run in main October repo

LukeTowers commented on .github/workflows/code-quality-pr.yaml:
```diff
@@ -10,13 +10,10 @@ jobs:
-          php-version: 7.2
-      - name: Install Composer dependencies
-        run: composer install --no-interaction --no-progress --no-suggest
+          php-version: '7.3'
```
What's with this change? Do our tests still support 7.2?

bennothommo replied:
Yeah, this is just for the Code Quality tests which run externally. It has no effect on the unit tests.

LukeTowers commented on src/Support/helpers.php:
```diff
@@ -210,13 +227,31 @@ function temp_path($path = '')
     }
 }
 
+if (!function_exists('e')) {
```
Did Laravel get rid of this helper and now we have to implement it ourselves?

bennothommo replied:
No, they didn't remove it, but they set the `$doubleEncode` parameter to `true` by default in 5.6 which stuffed up a lot of October functionality, so I opted to define the helper ourselves with the original `false` default.

bennothommo replied:
https://github.com/octobercms/library/commit/adda4366544e9e690b1a2b6d0cf3011df8a9ca3a

bennothommo committed:
Add config value to control loading of discovered Laravel packages (#492)

Adds the config value 'app.loadDiscoveredPackages' (default: false) - if true, packages discovered through Laravel's autodiscovery feature will be loaded. The reason for this control is that in some cases, packages which are autodiscovered belong to plugins. When the plugins themselves are disabled, the discovered package is still being loaded. See https://github.com/rainlab/debugbar-plugin/issues/34 for an example.

Additionally, in certain cases (for example, with Composer), sites will break when a package removed by Composer is loaded. This is because discovered packages and services are stored in cache ("storage/framework/services.php" and "storage/framework/packages.php").

This change will attempt to use the normal mechanism to load packages and services, and if it fails for any reason, it will delete the package and service cache and attempt to re-load the packages and services one more time (emulating someone running "php artisan config:clear" before "php artisan package:discover"). If a failure still occurs, then the failure is made known to the user.

Refs: https://github.com/octobercms/library/pull/492#issuecomment-636434950

LukeTowers committed:
Remove L6 branch specific build script
